encoding/thrift/inbound.go

```diff
@@ -62,6 +62,46 @@ func (t thriftUnaryHandler) Handle(ctx context.Context, treq *transport.Request,
 		return err
 	}
 
+	reader := bytes.NewReader(buf.Bytes())
+
+	if proto, ok := t.Protocol.(protocol.EnvelopeAgnosticProtocol); ok {
+
+		reqValue, responder, err := proto.DecodeRequest(reader)
+		if err != nil {
+			return err
+		}
+
+		res, err := t.UnaryHandler(ctx, reqValue)
+		if err != nil {
+			return err
+		}
+
+		if resType := res.Body.EnvelopeType(); resType != wire.Reply {
+			return errors.ResponseBodyEncodeError(
+				treq, errUnexpectedEnvelopeType(resType))
+		}
+
+		value, err := res.Body.ToWire()
+		if err != nil {
+			return err
+		}
+
+		if res.IsApplicationError {
+			rw.SetApplicationError()
```

Comment:
Instead of copying this logic, can we perhaps make our own responders for the
existing cases?

```go
var (
    reqValue wire.Value
    responder protocol.EnvelopeSpecificResponder
)
if proto, ok := t.Protocol.(protocol.EnvelopeAgnosticProtocol); ok {
    var err error
    reqValue, responder, err = proto.DecodeRequest(reader)
    if err != nil {
        return err
    }
} else if t.Enveloping {
    proto := t.Protocol
    envelope, err := proto.DecodeEnveloped(reader)
    if err != nil {
        return err
    }
    if envelope.Type != wire.Call {
        ...
    }
    reqValue = envelope.Value
    responder = envelopeResponder{SeqID: envelope.SeqID, Protocol: proto}
} else {
    proto := t.Protocol
    var err error
    reqValue, err = proto.Decode(reader, wire.TStruct)
    if err != nil {
        return err
    }
    responder = unenvelopedResponder{Protocol: proto}
}

// ...
```

Where `envelopeResponder` is an `EnvelopeSpecificResponder` that always relies
on `EncodeEnveloped` and `unenvelopedResponder` is one that always relies on
`Encode`.

Reply:
+1 to this.

Generally, though, I'm astounded by how surgical this change is.

Reply:
@abhinav What do you think of making the responders I’ve added to thriftrw public to this end?

Reply:
Preliminary investigation into this is disappointing. Doesn’t seem to be possible to do this and preserve the mocked proto tests, since they depend on the implementation to call through itself to encode the response.